### PR TITLE
WIP: Meta-data persistence and "type" field for node. 

### LINKED ISSUE #8: Add 'type': string as data point to node and relationship models.
Scope:
- add type to node and relationship
- update metadata: rename 'type' to 'modelName' (?). Add 'type?' (optional)
- update metadata generation logic for the above
- update metadata storage logic for the above as follows:
If metadata has a 'type' field then the file path would be: ~/metadata/{modelName}/type/{type.dat} with content as the JSON array of all the IDs of the node or relationships for that specific type.

### PR DESCRIPTION
**Pull request by Pronoy999**
Closes #8 
Added Metadata persistence storage and and added "type" for node. 

@dawndebanik How to add 2 tests for same method call? The first one is expecting the data for the second one. 

Signed-off-by: Pronoy Mukherjee <mukherjee_pronoy@yahoo.in>

### CODE REVIEW

File: src/models/node.ts [hunk 1]
```diff
@@ -1,10 +1,11 @@
-import { Models } from "../constants";
-import MetadataGenerable, { Metadata } from "../metadata/MetadataGenerable";
+import {Models} from "../constants";
+import MetadataGenerable, {Metadata} from "../metadata/MetadataGenerable";
 import Relationship from "./relationship";
 
 export default class Node implements MetadataGenerable {
   constructor(
     readonly id: number,
+    private type?: string,
```
Comment (dawndebanik) [inline src/models/node.ts]: Why is this optional? All nodes have a type. Mandatory Field.

File: src/models/relationship.ts [hunk 2]
```diff
@@ -1,5 +1,5 @@
-import { Models } from "../constants";
-import MetadataGenerable, { Metadata } from "../metadata/MetadataGenerable";
+import {Models} from "../constants";
+import MetadataGenerable, {Metadata} from "../metadata/MetadataGenerable";
 import Node from "./node";
 
 export default class Relationship implements MetadataGenerable {
```
Comment (dawndebanik) [inline src/models/relationship.ts:5]: Same as node. Mandatory type.

File: src/models/node.ts (hunk 1 above)
Comment (Pronoy999) [reply src/models/node.ts]: A user might choose not to specify a type.

**Comment by dawndebanik** [comment]
If you need any help on this, let me know.

**Comment by Pronoy999** [comment]
> If you need any help on this, let me know.

Actually, I can't mock that a file exists, that's why the coverage is not 100%. I checking other options.

File: src/persistance/metadata-writer.ts [hunk 3]
```diff
@@ -1,13 +1,30 @@
...
+
 export default class MetadataWriter {
   writeMetadata(metaData: Metadata): void {
-    const fileName: string = metaData.id + ".dat";
-    const fileLocation =
-      process.env[META_DATA_PATH_KEY] + metaData.type + "/" + fileName;
+    const metadataRootDir = process.env[META_DATA_PATH_KEY];
 
-    fs.writeFileSync(fileLocation, JSON.stringify(metaData.data), "utf-8");
+    const typeFileName = `${metaData.type}.${DATA_FILE_EXTENSION}`;
```
Comment (dawndebanik) [inline src/persistance/metadata-writer.ts:11]: Why did you remove the `if-else` check I had done on type? This `metadata.type` is an optional '?' field which can be undefined. This same writer is used by database and graph models, which doesn't have a type. It would write to `undefined.dat` for those cases.
